dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/search/SolrQuerySanitizer.java

```diff
@@ -0,0 +1,143 @@
+package edu.harvard.iq.dataverse.search;
+
+import com.google.common.collect.Sets;
+import edu.harvard.iq.dataverse.DatasetFieldServiceBean;
+import edu.harvard.iq.dataverse.persistence.dataset.DatasetFieldType;
+import org.apache.commons.lang.StringUtils;
+
+import javax.ejb.Stateless;
+import javax.inject.Inject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Stateless
+public class SolrQuerySanitizer {
+
+    private final static Set<Character> SOLR_SPECIAL_CHARACTERS = Sets.newHashSet(
+            '\\', ':', '+', '-', '!', '(', ')', 
+            '^' , '[', ']', '"', '{', '}', '~',
+            '*' , '?', '|', '&', ';', '/');
+    
+    
+    @Inject
+    private DatasetFieldServiceBean datasetFieldService;
+    @Inject
+    private SolrFieldFactory solrFieldFactory;
+    
+    // -------------------- LOGIC --------------------
+    
+    public String sanitizeQuery(String query) {
+        if (query == null) {
+            return StringUtils.EMPTY;
+        }
+        
+        query = replaceQueryFieldNames(query, buildFieldNamesMapping());
+        query = escapeGlobalIdValues(query);
+        
+        return query;
+    }
+    
+    // -------------------- PRIVATE --------------------
+    
+    private Map<String, String> buildFieldNamesMapping() {
+        Map<String, String> fieldNamesMapping = new HashMap<>();
+        List<DatasetFieldType> datasetFields = datasetFieldService.findAllOrderedById();
+        for (DatasetFieldType datasetFieldType : datasetFields) {
+
+            SolrField dsfSolrField = solrFieldFactory.getSolrField(datasetFieldType.getName(),
+                                                                   datasetFieldType.getFieldType(),
+                                                                   datasetFieldType.isThisOrParentAllowsMultipleValues(),
+                                                                   datasetFieldType.isFacetable());
+            fieldNamesMapping.put(datasetFieldType.getName(), dsfSolrField.getNameSearchable());
+        }
+        return fieldNamesMapping;
+    }
+    
+    private void transferString(StringBuilder source, StringBuilder target) {
+        target.append(source);
+        source.setLength(0);
+    }
+    
+    private String replaceQueryFieldNames(String query, Map<String, String> fieldNamesMapping) {
```

Comment:
After giving it some thought, I think this is good place for javadoc explaining the process.

Reply:
You mean how method is implemented? I'm not sure if this wouldn't be more confusing then it would explain things. I could add some description what it does, why it is done and maybe some example.
As for implementation details it probably will be better to have side comments inside method or nothing added at all (I can't think how to write it that will add some value)

Reply:
Hmm, I guess some description would do, I would avoid comments. Private methods with javadocs would be better but I don't think they would help in this case.